src/ess/imaging/normalize.py

```diff
@@ -0,0 +1,262 @@
+
+import warnings
+from typing import NewType
+
+import scipp as sc
+
+from .io import (
+    TIME_COORD_NAME,
+    DarkCurrentImageStacks,
+    OpenBeamImageStacks,
+    SampleImageStacks,
+)
+
+AverageBackgroundPixelCounts = NewType("AverageBackgroundPixelCounts", sc.Variable)
+"""~math:`D0 = mean(background)`."""
+AverageSamplePixelCounts = NewType("AverageSamplePixelCounts", sc.Variable)
+"""~math:`D = mean(sample)`."""
+ScaleFactor = NewType("ScaleFactor", sc.Variable)
+"""~math:`ScaleFactor = AverageBackgroundPixelCounts / AverageSamplePixelCounts`."""
+
+OpenBeamImage = NewType("OpenBeamImage", sc.DataArray)
+"""Open beam image. ~math:`average(open_beam)`."""
+DarkCurrentImage = NewType("DarkCurrentImage", sc.DataArray)
+"""Dark current image ~math:`average(dark_current)`."""
+AveragedNonSampleImages = NewType("AveragedNonSampleImages", sc.DataArray)
+"""Averaged open beam and dark current images."""
+BackgroundImage = NewType("BackgroundImage", sc.DataArray)
+"""Background image stack. ~math:`background = open_beam - dark_current`."""
+CleansedSampleImages = NewType("CleansedSampleImages", sc.DataArray)
+"""Sample image stack - dark current."""
+NormalizedSampleImages = NewType("NormalizedSampleImages", sc.DataArray)
+"""Normalized sample image stack.
+
+~math:`normalized_sample = sample / background * DFactor`.
+"""
+GroupedSampleImages = NewType("GroupedSampleImages", sc.DataArray)
+"""Grouped sample image stack by rotation angle."""
+
+
+def _mean_all_dims(data: sc.Variable) -> sc.Variable:
+    """Calculate the mean of all dimensions one by one to avoid overflow."""
+    if data.shape == ():  # scalar
+        return data
+    return _mean_all_dims(data.mean(dim=data.dims[0]))
+
+
+def average_open_beam_images(open_beam: OpenBeamImageStacks) -> OpenBeamImage:
+    """Average the open beam image stack.
+
+    .. math::
+
+        OpenBeam = mean(open_beam, 'time')
+
+    """
+    warnings.warn(
+        "Calculating average open beam image assuming constant exposure time.",
+        stacklevel=0,
+    )
+    return OpenBeamImage(sc.mean(open_beam, dim=TIME_COORD_NAME))
+
+
+def average_dark_current_images(
+    dark_current: DarkCurrentImageStacks,
+) -> DarkCurrentImage:
+    """Average the dark current image stack.
+
+    .. math::
+
+        DarkCurrent = mean(dark_current, 'time')
+
+    """
+    warnings.warn(
+        "Calculating average dark current image assuming constant exposure time.",
+        stacklevel=0,
+    )
+    return DarkCurrentImage(sc.mean(dark_current, dim=TIME_COORD_NAME))
+
+
+BackgroundPixelThreshold = NewType("BackgroundPixelThreshold", sc.Variable)
+"""Threshold for the background pixel values."""
+
+DEFAULT_BACKGROUND_THRESHOLD = BackgroundPixelThreshold(sc.scalar(1.0, unit="counts"))
+
+
+def calculate_white_beam_background(
+    open_beam: OpenBeamImage,
+    dark_current: DarkCurrentImage,
+    background_threshold: BackgroundPixelThreshold = DEFAULT_BACKGROUND_THRESHOLD,
```

Comment:
I'm thinking that `BackgroundPixelThreshold` should not have a default value here, but instead a default param should be set on the workflow/pipeline ?

Reply:
It does have a default param in the workflow/pipeline.

The default value in the function signature is not used by the `sciline.Pipeline`.

I thought it'll make it easier to use the provider itself, i.e. for testing...

Reply:
I would say if you want to use it for testing, then it's good that the parameter is required, so that you understand exactly what is being done?

An alternative would be to split it up into 2 providers: one that does the subtraction, and one that applies the threshold afterwards, with a very obvious name like `ReplaceZerosAndNegativesWithOne`?

Reply:
> I would say if you want to use it for testing, then it's good that the parameter is required, so that you understand exactly what is being done?

You can still test it with the default argument value though.

> An alternative would be to split it up into 2 providers: one that does the subtraction, and one that applies the threshold afterwards, with a very obvious name like ReplaceZerosAndNegativesWithOne?

I didn't want to hard-code the numbers so the name should be different but
I can split this up into two steps if that makes it more interpretable.

Reply:
Second thoughts... I think rounding up should be done within the normalization step since it's actually where it's needed.

Reply:
@nvaytet  I ended up inserting threshold applying steps instead of merging them into the normalization step
since they are used for calculating scale factor as well...

Reply:
But I still don't see why we can't use the default argument value though

Reply:
I'm saying that because I think in `esssans`, we don't have default arguments in providers, only default parameters on the workflow (unless I missed something somewhere).
So I thought we should try to be consistent across techniques.